File: androidApp/app/src/main/java/edu/uchicago/skgrogg/movies/screens/authenticate/AuthActivity.java
package edu.uchicago.skgrogg.movies.screens.authenticate;

import android.content.Intent;
import android.os.Bundle;
import android.util.Log;

import androidx.appcompat.app.AppCompatActivity;

import com.amazonaws.mobile.client.AWSMobileClient;
import com.amazonaws.mobile.client.Callback;
import com.amazonaws.mobile.client.SignInUIOptions;
import com.amazonaws.mobile.client.UserStateDetails;
import com.google.gson.Gson;

import edu.uchicago.skgrogg.movies.MainActivity;
import edu.uchicago.skgrogg.movies.cache.Cache;
import edu.uchicago.skgrogg.movies.models.Token;
import edu.uchicago.skgrogg.movies.common.utils.JWTUtils;
import edu.uchicago.skgrogg.movies.R;

public class AuthActivity extends AppCompatActivity {

    //used as the default email in the event authentication is disabled
    public static final String DEFAULT_COM = "default@default.com";
    private final String TAG = AuthActivity.class.getSimpleName();
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        //if use_auth is set to false, then short circuit
        String strUseAuth =  getResources().getString(R.string.use_auth);
        if (!Boolean.parseBoolean(strUseAuth)){
            Cache.getInstance().setUserEmail(DEFAULT_COM);
            Intent intent = new Intent(AuthActivity.this, MainActivity.class);
            startActivity(intent);

        } else {
            AWSMobileClient.getInstance().initialize(getApplicationContext(), new Callback<UserStateDetails>() {
                @Override
                public void onResult(UserStateDetails userStateDetails) {
                    Log.d(TAG, userStateDetails.getUserState().toString());
                    switch (userStateDetails.getUserState()){
                        case SIGNED_IN:
                            String strCodedToken =  userStateDetails.getDetails().get("token");
                            setEmail(strCodedToken);
                            Intent intent = new Intent(AuthActivity.this, MainActivity.class);
                            startActivity(intent);
                            break;
                        case SIGNED_OUT:
                            showSignIn();
                            break;
                        default:
                            AWSMobileClient.getInstance().signOut();
                            showSignIn();
                            break;

                    }
                }
                @Override
                public void onError(Exception e){
                    Log.e(TAG, e.toString());
                }
            });
        }


    }


    private void setEmail(String codedToken) {
        String decoded = "";
        try {
            decoded =  JWTUtils.decodedBody(codedToken);
            Gson gson = new Gson();
            Token token = gson.fromJson(decoded, Token.class);
            Cache.getInstance().setUserEmail(token.getEmail());
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    private void showSignIn() {
        try {
            AWSMobileClient.getInstance().showSignIn(this,
                    SignInUIOptions.builder()
                            .nextActivity(MainActivity.class)
                            .build(),
            new Callback<UserStateDetails>() {
                @Override
                public void onResult(UserStateDetails result) {
                    Log.d(TAG, "showSignIn() onResult() result: userState: " + result.getUserState());
                    switch (result.getUserState()){
                        case SIGNED_IN:
                            Log.d(TAG, "showSignIn() callback: SIGNED_IN logged in!");
                            setEmail(result.getDetails().get("token"));
                            break;
                        case SIGNED_OUT:
                            Log.d(TAG, "showSignIn() callback onResult: SIGNED_OUT ");
                            break;
                        default:
                            Log.d(TAG, "showSignIn() callback onResult: default; Should not be possible.");
                            break;
                    }
                }

                @Override
                public void onError(Exception e) {
                    Log.e(TAG, "showSignIn().onError: ", e);
                }
            }

            );


        } catch (Exception e) {
            Log.e("AuthActivity", e.toString());
        }
    }
}

File: androidApp/app/src/main/java/edu/uchicago/skgrogg/movies/screens/authenticate/AuthActivity.kt
package edu.uchicago.skgrogg.movies.screens.authenticate

import android.content.Intent
import android.os.Bundle
import android.util.Log
import androidx.appcompat.app.AppCompatActivity
import com.amazonaws.mobile.client.*
import com.google.gson.Gson
import edu.uchicago.skgrogg.movies.MainActivity
import edu.uchicago.skgrogg.movies.cache.Cache
import edu.uchicago.skgrogg.movies.common.Constants
import edu.uchicago.skgrogg.movies.common.utils.JWTUtils
import edu.uchicago.skgrogg.movies.models.Token
import java.lang.Boolean
import kotlin.Exception
import kotlin.String

class AuthActivity: AppCompatActivity() {

    //used as the default email in the event authentication is disabled
    val DEFAULT_COM = "default@default.com"
    val TAG = AuthActivity::class.java.simpleName
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        //if use_auth is set to false, then short circuit
        val strUseAuth: String = "false"
        if (!Boolean.parseBoolean(strUseAuth)) {
            Log.d("Cognito Effort", DEFAULT_COM)
            Cache.getInstance().userEmail = DEFAULT_COM
            val intent = Intent(this@AuthActivity, MainActivity::class.java)
            startActivity(intent)
        } else {
            AWSMobileClient.getInstance()
                .initialize(getApplicationContext(), object : Callback<UserStateDetails> {
                    override fun onResult(userStateDetails: UserStateDetails) {
                        Log.d(TAG, userStateDetails.userState.toString())
                        when (userStateDetails.userState) {
                            UserState.SIGNED_IN -> {
                                val strCodedToken = userStateDetails.details["token"]
                                setEmail(strCodedToken)
                                val intent = Intent(this@AuthActivity, MainActivity::class.java)
                                startActivity(intent)
                            }
                            UserState.SIGNED_OUT -> showSignIn()
                            else -> {
                                AWSMobileClient.getInstance().signOut()
                                showSignIn()
                            }
                        }
                    }

                    override fun onError(e: Exception) {
                        Log.e(TAG, e.toString())
                    }
                })
        }
    }


    open fun setEmail(codedToken: String?) {
        var decoded: String? = ""
        try {
            decoded = JWTUtils.decodedBody(codedToken)
            val gson = Gson()
            val token = gson.fromJson(
                decoded,
                Token::class.java
            )
            Cache.getInstance().userEmail = token.email
            Constants.userEmial = token.email
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    open fun showSignIn() {
        try {
            AWSMobileClient.getInstance().showSignIn(this,
                SignInUIOptions.builder()
                    .nextActivity(MainActivity::class.java)
                    .build(),
                object : Callback<UserStateDetails> {
                    override fun onResult(result: UserStateDetails) {
                        Log.d(TAG, "showSignIn() onResult() result: userState: " + result.userState)
                        when (result.userState) {
                            UserState.SIGNED_IN -> {
                                Log.d(TAG, "showSignIn() callback: SIGNED_IN logged in!")
                                setEmail(result.details["token"])
                            }
                            UserState.SIGNED_OUT -> Log.d(
                                TAG,
                                "showSignIn() callback onResult: SIGNED_OUT "
                            )
                            else -> Log.d(
                                TAG,
                                "showSignIn() callback onResult: default; Should not be possible."
                            )
                        }
                    }

                    override fun onError(e: Exception) {
                        Log.e(TAG, "showSignIn().onError: ", e)
                    }
                }
            )
        } catch (e: Exception) {
            Log.e("AuthActivity", e.toString())
        }
    }
}
